libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -265,7 +265,120 @@ string ABIFunctions::cleanupFunction(Type const& _type, bool _revertOnFailure)
 				StateMutability::Payable :
 				StateMutability::NonPayable
 			);
-			templ("body", "cleaned := " + cleanupFunction(addressType, _revertOnFailure) + "(value)");
+			templ("body", "cleaned := " + cleanupFunction(addressType) + "(value)");
+			break;
+		}
+		case Type::Category::Enum:
+		{
+			// Out of range enums cannot be truncated unambigiously and therefore it should be an error.
+			templ("body", "cleaned := " + validatorFunction(_type) + "(value)");
+			break;
+		}
+		case Type::Category::InaccessibleDynamic:
+			templ("body", "cleaned := 0");
+			break;
+		default:
+			solAssert(false, "Cleanup of type " + _type.identifier() + " requested.");
+		}
+
+		return templ.render();
+	});
+}
+
+string ABIFunctions::validatorFunction(Type const& _type, bool _revertOnFailure)
+{
+	string functionName = string("validator_") + (_revertOnFailure ? "revert_" : "assert_") + _type.identifier();
+	return createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(value) -> cleaned {
+				<body>
+			}
+		)");
+		templ("functionName", functionName);
+		switch (_type.category())
+		{
+		case Type::Category::Address:
+			templ("body", "cleaned := " + validatorFunction(IntegerType(160), _revertOnFailure) + "(value)");
+			break;
+		case Type::Category::Integer:
+		{
+			IntegerType const& type = dynamic_cast<IntegerType const&>(_type);
+			if (type.numBits() == 256)
+				templ("body", "cleaned := value");
+			else if (type.isSigned())
+			{
+				Whiskers w("if iszero(eq(value, signextend(<signbit>, value))) { <failure> } cleaned := value");
```

Comment:
I guess this doesn't cover the case when there are dirty bits, but the sign bit is not set.

Reply:
signextend should work equally with a sign bit of zero and one.

Reply:
It "does work", but in case the sign bit is 0, it ignores the top bits entirely.

That means (some of) the top bits might be set, even though the sign bit (for a uint128 for example) may be 0.

Reply:
Oh wow, are you sure? signextend should extend the sign bit no matter whether it is zero or one.

Reply:
Right, I was speaking from experience looking at implementations, and it seems the one I looked at was broken 😉